File: Java/src/Encoder.java
import sun.jvm.hotspot.debugger.cdbg.Sym;

import java.util.Arrays;

/**
 * Created by Oliver Poole(12022846) on 25/11/15.
 */
public class Encoder {

    static FrequencyTable frequencyTable = new FrequencyTable();

    public static double encode(String inputString) {

        int numberOfElements = inputString.length();

        // Sort the characters in the inputString
        char[] splitString = inputString.toCharArray();
        Arrays.sort(splitString);

        String sortedInputString = new String(splitString);

        double previousLower = 0.0;
        double probability;

        /**
         * Add the values to the frequency table
         */

        for (char character : sortedInputString.toCharArray()) {
            frequencyTable.addValue(character);
        }


        /**
         *  Generate the frequency table
         */

        for (Symbol symbol : frequencyTable.table) {

            probability = (double) symbol.frequency / numberOfElements;

            symbol.lowerRange = previousLower;
            symbol.upperRange = probability + previousLower;

            previousLower = probability + previousLower;
        }



        /**
         *  Encoding the input string
         */

        double lower = 0.0;
        double upper = 1.0;

        for (char inputCharacter: inputString.toCharArray()) {
            double range = upper - lower;

            Symbol symbol = frequencyTable.symbolForElement(inputCharacter);

            upper = lower + symbol.upperRange * range;
            lower = lower + symbol.lowerRange * range;
        }

        return lower;

    }
}

File: Swift/Arithmetic Coding/Encoder.swift
//
//  Encoder.swift
//  Arithmetic Coding
//
//  Created by Oliver Poole on 24/11/2015.
//  Copyright © 2015 OliverPoole. All rights reserved.
//

import Foundation

class Encoder {
    
    /**
     Encodes a string input to double form
     
     - parameter inputString: The string to encode
     
     - returns: A tuple containing the frequency table used in the convertion and the encoded output
     */
    static func encode(inputString : String) -> (table : FrequencyTable, encodedInput : Double) {
        
        /**
         *  Creating the frequency table
         */
        let frequencyTable = FrequencyTable()
        
        let numberOfElements = inputString.characters.count
        
        let sortedInput = inputString.characters.sort()
        
        
        for char in sortedInput {
            frequencyTable.addValue(char)
        }
        
        
        var previousLower : Double = 0.0
        var probability : Double = 0.0
        
        // Assign the frequencies of each input character
        for symbol in frequencyTable.table.values {
            
            probability = Double(symbol.frequency) / Double(numberOfElements)
            
            // set the range for each symbol
            symbol.range = SymbolRange(lower: previousLower, upper: (probability + previousLower))
            
            // Set the previous lower for the next symbol
            previousLower = probability + previousLower
        }
        
        
        
        /**
         *  Encoding the input string
         */
        var low : Double = 0.0
        var high : Double = 1.0
        
        for inputSymbol in inputString.characters {
            
            // Calculate the range
            let range = high - low
            
            // Get the symbol from the table
            let symbol = frequencyTable.symbolForElement(inputSymbol)!
            
            // Calculate the new high and low values
            high = low + (symbol.range!.upper * range)
            low = low + (symbol.range!.lower * range)
            
        }
        
        // Return the frequency table, and the average of the low and high values
        return (frequencyTable, ((low + high) / 2.0))
    }
}